File: uis/application/lib/incidents.ts
export const incidentCategories = [
  "RETURN_REQUEST",
  "DAMAGE",
  "DELAYED_DELIVERY",
  "WRONG_ADDRESS",
  "LOST_PARCEL",
] as const;

export const incidentStatuses = ["open", "in_progress", "resolved", "discarded"] as const;
export const incidentOrigins = ["customer", "branch", "internal"] as const;
export const incidentBranches = ["Los Angeles", "Zaragoza", "central"] as const;

export type IncidentCategory = (typeof incidentCategories)[number];
export type IncidentStatus = (typeof incidentStatuses)[number];
export type IncidentOrigin = (typeof incidentOrigins)[number];
export type IncidentBranch = (typeof incidentBranches)[number];

export type Incident = {
  id: string;
  title: string;
  description: string;
  category: IncidentCategory;
  status: IncidentStatus;
  origin: IncidentOrigin;
  branch: IncidentBranch;
  created_at: string;
  updated_at: string;
};

export type IncidentSummary = {
  total: number;
  by_status: Record<IncidentStatus, number>;
  by_category: Record<IncidentCategory, number>;
  by_origin: Record<IncidentOrigin, number>;
  by_branch: Record<IncidentBranch, number>;
};

const apiBase = "/backend/api/incidents";

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await fetch(`${apiBase}${path}`, { ...init, cache: "no-store" });
  if (!response.ok) {
    throw new Error(response.status >= 500 ? "El servicio no está disponible ahora." : "Revisa los datos introducidos.");
  }
  return (await response.json()) as T;
}

export function listIncidents(filters: { status?: string; origin?: string; branch?: string }) {
  const params = new URLSearchParams();
  for (const [key, value] of Object.entries(filters)) {
    if (value) params.set(key, value);
  }
  const query = params.toString();
  return request<Incident[]>(query ? `?${query}` : "");
}

export function getIncidentSummary() {
  return request<IncidentSummary>("/summary");
}

export function createIncident(payload: Omit<Incident, "id" | "created_at" | "updated_at">) {
  return request<Incident>("", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
}

export function updateIncidentStatus(id: string, status: IncidentStatus) {
  return request<Incident>(`/${id}/status`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ status }),
  });
}

File: services/api/routes/incidents.py
from typing import List, Optional

from fastapi import APIRouter, Query

from ..models import (
    Incident,
    IncidentCategory,
    IncidentCreate,
    IncidentOrigin,
    IncidentStatus,
    IncidentStatusUpdate,
    IncidentSummary,
)
from ..services.incidents import (
    create_incident,
    get_incident,
    list_incidents,
    summary,
    update_incident_status,
)

router = APIRouter(prefix="/api/incidents", tags=["Incidents"])


@router.post("", response_model=Incident, status_code=201)
def post_incident(payload: IncidentCreate):
    return create_incident(payload)


@router.get("", response_model=List[Incident])
def get_incidents(
    status: Optional[IncidentStatus] = None,
    origin: Optional[IncidentOrigin] = None,
    branch: Optional[str] = Query(default=None),
    category: Optional[IncidentCategory] = None,
):
    return list_incidents(status, origin, branch, category)


@router.get("/summary", response_model=IncidentSummary)
def get_incidents_summary():
    return summary()


@router.get("/{incident_id}", response_model=Incident)
def get_incident_detail(incident_id: int):
    return get_incident(incident_id)


@router.patch("/{incident_id}/status", response_model=Incident)
def patch_incident_status(incident_id: int, payload: IncidentStatusUpdate):
    return update_incident_status(incident_id, payload)
